### PR TITLE
Use new WinRT APIs for SpecialFolders

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Starting in master, then I'll port to UWP.

### CODE REVIEW

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln [hunk 1]
```diff
@@ -75,8 +78,8 @@ Global
 		{9F312D76-9AF1-4E90-B3B0-815A1EC6C346}.Debug|Any CPU.Build.0 = netstandard-Debug|Any CPU
 		{9F312D76-9AF1-4E90-B3B0-815A1EC6C346}.Release|Any CPU.ActiveCfg = netstandard-Release|Any CPU
 		{9F312D76-9AF1-4E90-B3B0-815A1EC6C346}.Release|Any CPU.Build.0 = netstandard-Release|Any CPU
-		{845D2B72-D8A4-42E5-9BE9-17639EC4FC1A}.Debug|Any CPU.ActiveCfg = netcoreapp-Windows_NT-Debug|Any CPU
-		{845D2B72-D8A4-42E5-9BE9-17639EC4FC1A}.Debug|Any CPU.Build.0 = netcoreapp-Windows_NT-Debug|Any CPU
+		{845D2B72-D8A4-42E5-9BE9-17639EC4FC1A}.Debug|Any CPU.ActiveCfg = uapaot-Windows_NT-Debug|Any CPU
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: run `msbuild /t:updateVSConfigurations` at the root?

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 2]
```diff
@@ -250,14 +255,24 @@
     </Compile>
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="..\..\System.Private.Uri\src\System.Private.Uri.csproj" />
-    <ProjectReference Include="..\..\System.Diagnostics.Debug\src\System.Diagnostics.Debug.csproj" />
-    <ProjectReference Include="..\..\System.Security.Principal\src\System.Security.Principal.csproj" />
-    <ProjectReference Include="..\..\System.Runtime\src\System.Runtime.csproj" />
+    <ProjectReference Include="..\..\System.Private.Uri\src\System.Private.Uri.csproj">
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: @joperezr  or @weshaggard should comment on this. I thought our build was set up to "do the right thing". I don't see other cases where projects set this manually.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            // For testing we'll fall back if the needed APIs aren't present
+            if (ApiInformation.IsTypePresent("Windows.Storage.UserDataPaths"))
+            {
+                return GetFolderPathCoreCurrent(folder, option);
+            }
+            else
+            {
+                return GetFolderPathCoreFallBack(folder, option);
+            }
+
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: nit newline here and 28

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        private static string GetFolderPathCoreFallBack(SpecialFolder folder, SpecialFolderOption option)
+        {
+            switch (folder)
+            {
+                case SpecialFolder.ApplicationData:
+                    return ApplicationData.Current.RoamingFolder?.Path;
+                case SpecialFolder.LocalApplicationData:
+                    return ApplicationData.Current.LocalFolder?.Path;
+                default:
```
Comment (danmosemsft) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: what about SpecialFodler.CommonApplicationData --> ApplicationData.Current.SharedLocalFolder?.Path;

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Rather start tight and expand where needed. This is for internal use only.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: This is just a hack to use while in VS. Was trying to save myself a little time the next time I needed to hack up the project.

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: Whoops, I'll fix this one.

**Comment by danmosemsft** [comment]
Are you going to add tests? Worst case, they can validate that the result is either empty or the value we expect. That at least verifies the code path doesn't crash. Better case, the test checks the Windows build number, and if it's high enough, actually runs.

**Comment by JeremyKuhne** [comment]
> Are you going to add tests? 

The tests are already there. Nothing special for WinRT.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 2 above)
Comment (weshaggard) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: We should not be passing in the AdditionalProperties here as that will definitely fight against our configuration system.

**Comment by danmosemsft** [comment]
>The tests are already there. Nothing special for WinRT.

Right, but they're disabled. You need to remove `        [ActiveIssue(20782, TargetFrameworkMonikers.Uap)]`

Then we should make sure that on OS builds that actually have this API, they actually use it.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+using Windows.Storage;
+
+namespace System
+{
+    public static partial class Environment
+    {
+        private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOption option)
+        {
+            // For testing we'll fall back if the needed APIs aren't present
+            if (ApiInformation.IsTypePresent("Windows.Storage.UserDataPaths"))
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Is this check expensive?  I'm wondering if its result should be cached.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                    return SystemDataPaths.GetDefault().PublicVideos;
+                case SpecialFolder.UserProfile:
+                    return UserDataPaths.GetDefault().Profile;
+                case SpecialFolder.SystemX86:
+                    return SystemDataPaths.GetDefault().SystemX86;
+                case SpecialFolder.Windows:
+                    return SystemDataPaths.GetDefault().Windows;
+
+                // The following aren't available on WinRT. Our default behavior for
+                // is string.Empty for paths that aren't available.
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Nit: "for is"? Seems like a word is missing.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                return GetFolderPathCoreCurrent(folder, option);
+            }
+            else
+            {
+                return GetFolderPathCoreFallBack(folder, option);
+            }
+
+        }
+
+        private static string GetFolderPathCoreCurrent(SpecialFolder folder, SpecialFolderOption option)
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Is it correct that we're ignoring `option`, both here and in the fallback?  We implement that manually in the Unix implementation, for example:
https://github.com/dotnet/corefx/blob/475c578fc77e16344679e337276c91a70fe1c79c/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L67
https://github.com/dotnet/corefx/blob/475c578fc77e16344679e337276c91a70fe1c79c/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L89

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: I left it out on purpose. Didn't seem like a good thing to be trying to create folders in lowbox- worried about trying to do this and negatively impacting the system (bad ACLs, what have you). They (Windows) explicitly didn't expose the option in their new APIs.

I'll add comments.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 6 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Whoops- thanks

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: It didn't appear to be. I wouldn't expect it to be too bad given that it is a "normal" scenario (light up). I also wouldn't expect GetFolderPath to be in tight or critical loops.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: What about existence verification?

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: They actually implicitly turn it off. I'll call that out too.

**Comment by JeremyKuhne** [comment]
> Right, but they're disabled. 

Yes, sorry I wasn't clear. I'm still validating the existing tests on a _public_ build where the functionality exists. Want to make sure the conditions are correct for the _latest_ fast track RS3 build.  I'll have that complete by EOD tomorrow hopefully.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: I'm not understanding the rationale.  Our goal here is to achieve as much fidelity as possible with the existing .NET API, not with the underlying API we build on; if the underlying API doesn't provide the functionality but we can achieve it in other ways, why shouldn't we?  The default behavior is to verify that the path being handed back actually exists, and if it doesn't, give back an empty string so that consuming code doesn't try to use something that doesn't exist; by not doing that, it seems like we're artificially harming consuming code. Is there some reason we can't simply do the exists check?  As for Create, I understand the concerns, but code has to explicitly opt-in by passing in Create; if they do so, shouldn't we respect that?  If we don't, a) existing code that expected it won't work correctly and b) such code would then potentially be modified to work around it by doing the exact creation option we'd attempt.

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: On Windows the API creates a very specific ACL for folders through the API. It is asking for trouble to just call CreateDirectory- I'm super hesitant to try laying down a folder with potentially compat-breaking ACLs.

That said, I'm digging into the behavior a bit more. I've just set up another clean install- I'll circle back with more details. (It may be moot as the folders that actually are exposed are all standard folders...)

Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: Thanks.

**Comment by danmosemsft** [comment]
Please also delete these two files now:
```
C:\git\corefx\src\System.Runtime.Extensions\src\PinvokeAnalyzerExceptionList.analyzerdata.uap
C:\git\corefx\src\System.Runtime.Extensions\src\PinvokeAnalyzerExceptionList.analyzerdata.uapaot
```

File: dependencies.props [hunk 8]
```diff
@@ -157,7 +157,7 @@
     </StaticDependency>
 
     <StaticDependency Include="Microsoft.TargetingPack.Private.WinRT">
-      <Version>1.0.3</Version>
```
Comment (joperezr) [inline dependencies.props]: I believe this guy won't get used at all, so we can probably just remove it. cc: @dagood

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln (hunk 1 above)
Comment (joperezr) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: yeah, this one shouldn't be here as the active configuration, only on your local development.

---
In reply to: [129703878](https://github.com/dotnet/corefx/pull/22664#discussion_r129703878) [](ancestors = 129703878)

File: src/System.Runtime.Extensions/System.Runtime.Extensions.sln [hunk 9]
```diff
@@ -42,6 +42,9 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tests", "tests", "{1A2F9F4A-A032-433E-B914-ADD5992BB178}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{E107E9C1-E893-4E87-987E-04EF0DCEAEFD}"
+	ProjectSection(SolutionItems) = preProject
+		src\Configurations.props = src\Configurations.props
```
Comment (joperezr) [inline src/System.Runtime.Extensions/System.Runtime.Extensions.sln]: mm we don't have this file in any other solution, so it's weird to include it here.

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 10]
```diff
@@ -89,10 +89,21 @@
       <Link>Common\System\HResults.cs</Link>
     </Compile>
   </ItemGroup>
```
Comment (joperezr) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: The `TargetsWIndows` part of the condition is redundant, as uapaot and uap, will always be true

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 11]
```diff
@@ -89,10 +89,21 @@
...
+  <ItemGroup Condition="'$(TargetsWindows)' == 'true' and ('$(TargetGroup)' == 'uapaot' or '$(TargetGroup)' == 'uap')">
```
Comment (joperezr) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: The TargetsWIndows part of the condition is redundant, as uapaot and uap, will always be true

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj [hunk 12]
```diff
@@ -89,10 +89,21 @@
...
+    <Compile Include="System\Environment.WinRT.cs" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetsWindows)' == 'true' and '$(TargetGroup)' != 'uapaot' and '$(TargetGroup)' != 'uap'">
```
Comment (joperezr) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: We usually try to be inclusive in the conditions as opposed to exclusive. Can you try something like `TargetGroup == netcoreapp`

File: src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj (hunk 2 above)
Comment (joperezr) [inline src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj]: this was only meant to be a local hack to be able to build using VS, but shouldn't be checked in.

File: dependencies.props (hunk 8 above)
Comment (dagood) [inline dependencies.props]: A way to check is to give Version a garbage value (say, `9.9.9`) and run `msbuild /t:VerifyDependencies`. If there are no verification errors, then this item isn't doing anything useful.

**Comment by JeremyKuhne** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

File: dependencies.props (hunk 8 above)
Comment (JeremyKuhne) [inline dependencies.props]: I'll follow up on this in a separate change.

File: src/System.Runtime.Extensions/src/System/Environment.WinRT.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: > I'll circle back with more details

I see you merged but didn't see the results of this... what did you discover that made you decide not to address this?

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.WinRT.cs]: The comments were added to the code. Discussing offline with @ianhays and @danmosemsft we decided it was best to not do anything with the current data we have.

**Comment by JeremyKuhne** [comment]
> If they're inaccessible and the verify flag is set, isn't that a case where we should be handing back empty string?

Doesn't match legacy behavior. You could very easily not have rights (non-admin users and many locations) and Windows would still give the folder path back and we would pass it along.
